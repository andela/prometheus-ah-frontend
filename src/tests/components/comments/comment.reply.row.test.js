import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ConnectedCommentReplyRow,
{ CommentReplyRow } from '../../../components/comment/CommentReplyRow';
import mockData from '../../__mocks__/commentMockData';

const props = {
  deleteCommentAction: jest.fn(),
  updateCommentAction: jest.fn(),
  loadCommentsAction: jest.fn(),
  article: mockData.articles[0],
  comment: mockData.comments.comments[0],
  auth: true,
  error: {},
  username: 'faksam',
  reply: {
    id: 2,
    createdAt: '2016-02-18T03:22:56.637Z',
    updatedAt: '2016-02-18T03:48:35.824Z',
    body: 'This article is hot',
    User: {
      username: 'joeeasy',
      email: 'joeeasy@gmail.com',
      bio: 'tomorrow is a better day',
      id: 1,
      isVerified: true,
      role: 'user'
    }
  },
  commentReply: {
    id: 2,
    createdAt: '2016-02-18T03:22:56.637Z',
    updatedAt: '2016-02-18T03:48:35.824Z',
    body: 'This article is hot',
    User: {
      username: 'joeeasy',
      email: 'joeeasy@gmail.com',
      bio: 'tomorrow is a better day',
      id: 1,
      isVerified: true,
      role: 'user'
    }
  },
  commentReplies: [
    {
      id: 2,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
    {
      id: 3,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
    {
      id: 4,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
  ],
};
const initialState = {
  articleReducer: {
    article: mockData.articles[0],
  },
  commentReducer: {
    reply: mockData.commentReplies[0],
    commentReply: mockData.commentReplies[0],
    commentReplies: mockData.commentReplies,
  },
  auth: {
    error: {},
    user: {
      username: 'joeeasy',
      email: 'joeeasy@gmail.com',
      bio: 'tomorrow is a better day',
      id: 1,
      isVerified: true,
      role: 'user'
    }
  },
};
const mockStore = configureMockStore([thunk]);
const store = mockStore(initialState);

describe('CommentReplyRow Component', () => {
  it('should render without throwing an error', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it('should update a specific comment reply', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ editBody: 'Samuel is a boy' });
    instance.isValid();
    instance.onSubmit(event);
    expect(props.updateCommentAction).toHaveBeenCalled();
  });

  it('should not update a reply when reply body is empty', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({ editBody: '' });
    expect(instance.isValid()).toBe(false);
  });
  it('should delete a comment reply', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentDelete(event);
    expect(props.deleteCommentAction).toHaveBeenCalled();
  });
  it('should start editing a comment reply', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentEdit(event);
    expect(wrapper.state().editing).toEqual(true);
  });
  it('should cancel comment reply editing', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentEditCancel(event);
    expect(wrapper.state().editing).toEqual(false);
  });
  it('should set state to props', () => {
    const wrapper = shallow(<ConnectedCommentReplyRow store={store} {...props} />);
    expect(wrapper.state()).toEqual({});
  });

  it('should set reply body when the reply body changes', () => {
    const wrapper = shallow(<CommentReplyRow {...props} />);
    const event = {
      preventDefault: jest.fn(),
      target: {
        id: 'editBody',
        value: 'This is a comment'
      }
    };
    wrapper.setState({ errors: { [event.target.id]: 'mock' } });
    wrapper.instance().onChange(event);
    expect(wrapper.instance().state[event.target.id])
      .toBe(event.target.value);
  });
});
