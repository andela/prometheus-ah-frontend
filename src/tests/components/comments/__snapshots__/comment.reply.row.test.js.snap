// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentReplyRow Component should render without throwing an error 1`] = `
<div>
  <hr />
  <div
    className="row comment-reply-row"
  >
    <div
      className="col-2"
    >
      <img
        alt="user-thumbnail"
        className="rounded-circle float-left img img-fluid-comment"
        src={Object {}}
      />
    </div>
    <div
      className="col-10"
    >
      <div
        className="comment-body"
      >
        <p
          id="username"
        >
          <strong>
            joeeasy
          </strong>
          <span
            className="comment-span"
            id="timestamp"
          >
            -  
            3 years ago
          </span>
        </p>
        <p>
          This article is hot
        </p>
        <p
          className="comment-span"
        >
           
          <i
            className="mdi mdi-thumb-up-outline mr-3 align-top"
          >
            <span
              className="likes-num"
            >
               30
            </span>
          </i>
        </p>
      </div>
    </div>
    <div
      className="pull-right"
    />
    <br />
  </div>
</div>
`;
